### PR TITLE
UCP/UCT: add support fixed address for mmap

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
Fixes https://github.com/openucx/ucx/issues/1273
@shssf @brminich please review

### CODE REVIEW

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -247,11 +247,14 @@ enum ucp_dt_type {
...
  */
 enum {
-    UCP_MEM_MAP_NONBLOCK = UCS_BIT(0) /**< Complete the mapping faster, possibly by
+    UCP_MEM_MAP_NONBLOCK = UCS_BIT(0), /**< Complete the mapping faster, possibly by
                                            not populating the pages in the mapping
                                            up-front, and mapping them later when
                                            they are accessed by communication
                                            routines. */
+    UCP_MEM_MAP_FIXED    = UCS_BIT(1)  /**< Don't interpret addr as a hint:
+                                           place the mapping at exactly that
```
Comment (shssf) [inline src/ucp/api/ucp.h]: Comment indentation (in above lines also). Please look line 234 for example.

File: src/ucp/core/ucp_mm.c [hunk 2]
```diff
@@ -141,12 +141,15 @@ static ucs_status_t ucp_mem_alloc(ucp_context_h context, size_t length,
          * name specified in the configuration.
          */
         num_mds = 0;
-        if (method == UCT_ALLOC_METHOD_MD) {
+        if (method == UCT_ALLOC_METHOD_MD && !(uct_flags & UCT_MD_MEM_FLAG_FIXED)) {
```
Comment (shssf) [inline src/ucp/core/ucp_mm.c]: brackets should be used for both `if ((method == UCT_ALLOC_METHOD_MD) && !(uct_flags & UCT_MD_MEM_FLAG_FIXED))`

File: src/ucp/core/ucp_mm.c [hunk 3]
```diff
@@ -208,30 +228,31 @@ ucs_status_t ucp_mem_map(ucp_context_h context, ucp_mem_map_params_t *params,
         goto out;
     }
 
-    uct_flags = 0;
```
Comment (shssf) [inline src/ucp/core/ucp_mm.c]: Please use the function here `flags = uct_flags(params);`

File: src/ucp/core/ucp_mm.c [hunk 4]
```diff
@@ -175,11 +178,28 @@ static ucs_status_t ucp_mem_alloc(ucp_context_h context, size_t length,
...
+
+    return flags;
+}
+
 ucs_status_t ucp_mem_map(ucp_context_h context, ucp_mem_map_params_t *params,
                          ucp_mem_h *memh_p)
 {
     ucs_status_t status;
-    unsigned uct_flags;
+    unsigned flags = uct_flags(params);
```
Comment (shssf) [inline src/ucp/core/ucp_mm.c]: Please don't call the function `uct_flags()` here for several reasons:
1. it executes always despite of the fact the code below could avoid this execution
2. The return value might be wrong due to change in source data later.

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -376,6 +376,10 @@ enum {
                                                 mapping may be deferred until
                                                 it is accessed by the CPU or a
                                                 transport. */
+    UCT_MD_MEM_FLAG_FIXED    = UCS_BIT(1), /**< Don't interpret addr as a hint:
```
Comment (shssf) [inline src/uct/api/uct.h]: Please remove comma after `UCS_BIT(1)` because it reduce portability

File: src/ucp/core/ucp_mm.c (hunk 3 above)
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_mm.c]: Done, moved before first usage of `flags`

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3350/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/881/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2845/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3351/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/882/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2846/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3367/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/901/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/902/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2866/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3368/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3370/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3371/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2867/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/908/ for details.

**Comment by evgeny-leksikov** [comment]
See #1323

**Comment by yosefe** [comment]
pls close when PR is merged

**Pull request closed without merging**